### PR TITLE
EOS-14520 : Added UT's for delayed delete

### PR DESCRIPTION
**Pull request by pranavPawar07**
Signed-off-by: pranav <pranav.pawar@seagate.com>

### CODE REVIEW

File: ut/s3_delete_multiple_objects_action_test.cc [hunk 1]
```diff
@@ -845,3 +846,46 @@ TEST_F(S3DeleteMultipleObjectsResponseBodyTest, AllSucessQuietModeOff) {
...
+  S3Option::get_instance()->set_s3server_obj_delayed_del_enabled(true);
+
+  action_under_test->motr_kv_writer =
+      motr_kvs_writer_factory->mock_motr_kvs_writer;
+  action_under_test->motr_writer = motr_writer_factory->mock_motr_writer;
+
+  EXPECT_CALL(*(motr_writer_factory->mock_motr_writer),
+              delete_objects(_, _, _, _)).Times(0);
+
+  action_under_test->cleanup();
```
Comment (knrajnambiar76) [inline ut/s3_delete_multiple_objects_action_test.cc:861]: this testcase seems to be missing somethin, you are suppose to call the function to be tested. Here you are calling cleanup

File: ut/s3_delete_multiple_objects_action_test.cc [hunk 2]
```diff
@@ -845,3 +846,46 @@ TEST_F(S3DeleteMultipleObjectsResponseBodyTest, AllSucessQuietModeOff) {
...
+}
+
+TEST_F(S3DeleteMultipleObjectsActionTest, SizeBucketingOfObjects) {
```
Comment (knrajnambiar76) [inline ut/s3_delete_multiple_objects_action_test.cc:864]: you dont have to have this tests multiple times, if there is test file for Utilities then add to that, else in one place its good enough.

File: ut/s3_delete_object_action_test.cc [hunk 3]
```diff
@@ -289,6 +290,70 @@ TEST_F(S3DeleteObjectActionTest, SendAnyFailedResponse) {
...
+
+  action_under_test->delete_object();
+}
+
+TEST_F(S3DeleteObjectActionTest, DelayedDeleteObject) {
+  CREATE_OBJECT_METADATA;
+
+  S3Option::get_instance()->set_s3server_obj_delayed_del_enabled(true);
+
+  int layout_id = 2;
```
Comment (knrajnambiar76) [inline ut/s3_delete_object_action_test.cc:316]: you can remove this

File: ut/s3_delete_object_action_test.cc [hunk 4]
```diff
@@ -289,6 +290,70 @@ TEST_F(S3DeleteObjectActionTest, SendAnyFailedResponse) {
...
+  struct m0_uint128 obj_oid = {0x1ffff, 0x1ffff};
+  EXPECT_CALL(*(object_meta_factory->mock_object_metadata), get_oid())
+      .Times(AtLeast(1))
+      .WillRepeatedly(Return(obj_oid));
+  EXPECT_CALL(*(object_meta_factory->mock_object_metadata), get_layout_id())
+      .Times(0);
+  EXPECT_CALL(*(motr_writer_factory->mock_motr_writer),
+              delete_object(_, _, layout_id)).Times(0);
```
Comment (knrajnambiar76) [inline ut/s3_delete_object_action_test.cc:324]: instead of layout_id you can have _

File: ut/s3_put_chunk_upload_object_action_test.cc [hunk 5]
```diff
@@ -693,6 +694,58 @@ TEST_F(S3PutChunkUploadObjectActionTestNoAuth,
...
+  action_under_test->clear_tasks();
+  ACTION_TASK_ADD_OBJPTR(
+      action_under_test,
+      S3PutChunkUploadObjectActionTestBase::func_callback_one, this);
+
+  action_under_test->delete_old_object();
+  EXPECT_EQ(1, call_count_one);
+}
+
+TEST_F(S3PutChunkUploadObjectActionTestNoAuth, SizeBucketingOfObjects) {
```
Comment (knrajnambiar76) [inline ut/s3_put_chunk_upload_object_action_test.cc:720]: only one place enough

File: ut/s3_post_complete_action_test.cc [hunk 6]
```diff
@@ -668,23 +669,77 @@ TEST_F(S3PostCompleteActionTest, DeleteNewObject) {
...
+  EXPECT_CALL(*(motr_writer_factory->mock_motr_writer), set_oid(_))
+      .Times(AtLeast(1));
+  EXPECT_CALL(*(motr_writer_factory->mock_motr_writer),
+              delete_object(_, _, old_layout_id)).Times(AtLeast(1));
 
-//   EXPECT_CALL(*(motr_writer_factory->mock_motr_writer), set_oid(_))
-//       .Times(AtLeast(1));
-//   EXPECT_CALL(*(motr_writer_factory->mock_motr_writer),
-//               delete_object(_, _, old_layout_id)).Times(1);
+  action_under_test_ptr->delete_old_object();
```
Comment (AtitaShirwaikar) [inline ut/s3_post_complete_action_test.cc:689]: Remove commented code.Is this test case incomplete? Shouldnt we expect some action to happen after deleting the old object?

Comment (pranavPawar07) [reply ut/s3_post_complete_action_test.cc:689]: Commented code was the previous UT which was commented in Siddhi's patch. Uncommented one is the newly modified one.

File: ut/s3_delete_multiple_objects_action_test.cc (hunk 1 above)
Comment (pranavPawar07) [reply ut/s3_delete_multiple_objects_action_test.cc:861]: The logic for delayed delete is in the cleanup() function and there is no other explicit function named delete_old_objects() / delete_object().

File: ut/s3_post_complete_action_test.cc (hunk 6 above)
Comment (pranavPawar07) [reply ut/s3_post_complete_action_test.cc:689]: Same thing here as well.

**Review by AtitaShirwaikar — APPROVED** [review]
+1

**Review by knrajnambiar76 — APPROVED** [review]

**Comment by pranavPawar07** [comment]
Jenkins pre merge job : **Passed**
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/293/
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/292/

**Pull request merged**